pkg/adt/interval_tree.go

```diff
@@ -413,26 +434,62 @@ type IntervalVisitor func(n *IntervalValue) bool
 // Visit calls a visitor function on every tree node intersecting the given interval.
 // It will visit each interval [x, y) in ascending order sorted on x.
 func (ivt *IntervalTree) Visit(ivl Interval, ivv IntervalVisitor) {
-	ivt.root.visit(&ivl, func(n *intervalNode) bool { return ivv(&n.iv) })
+	ivt.root.visit(&ivl, ivt.nilNode, func(n *intervalNode) bool { return ivv(&n.iv) })
 }
 
 // find the exact node for a given interval
-func (ivt *IntervalTree) find(ivl Interval) (ret *intervalNode) {
-	f := func(n *intervalNode) bool {
-		if n.iv.Ivl != ivl {
-			return true
+func (ivt *IntervalTree) find(ivl Interval) *intervalNode {
+	x := ivt.root
+
+	ok := false
+	for x != ivt.nilNode {
+		if ivl.Begin.Compare(x.iv.Ivl.Begin) == 0 && ivl.End.Compare(x.iv.Ivl.End) == 0 {
+			ok = true
+			break
 		}
-		ret = n
-		return false
+
+		if ivl.End.Compare(x.max) > 0 {
+			ok = true
+			x = ivt.nilNode
+			break
+		}
+
+		if ivl.Begin.Compare(x.iv.Ivl.Begin) < 0 {
+			x = x.left
+			continue
+		}
+
+		if ivl.Begin.Compare(x.iv.Ivl.Begin) > 0 {
+			x = x.right
+			continue
+		}
+
+		// when ivt.Begin == x.iv.Ivl.Begin need visit all nodes x-subtree
+		break
+	}
+
+	if !ok {
+		ret := ivt.nilNode
+		f := func(n *intervalNode) bool {
+			if n.iv.Ivl != ivl {
+				return true
+			}
+			ret = n
+			return false
+		}
+
+		x.visit(&ivl, ivt.nilNode, f)
+
+		return ret
 	}
-	ivt.root.visit(&ivl, f)
```

Comment:
why not just call visit as before?

Reply:
the visit function maybe visit all nodes for find one node, the new find function use BST nature can prune, but when find left interval same, the must use visit function visit the subtree all nodes for find the node

Reply:
if this is an optimization, can we move this to a separate PR?

Reply:
I think it is just make it better, not optimization

Reply:
OK. Then can we not include this enhancement in this PR?

Reply:
OK, the change the find function just use visit